Reader /discover SEO: Enable SSR and add a localized meta description

jornp opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to pxLjZ-8N9-p2.

## Proposed Changes

This PR:

* Fixes i18n of `/discover` and adds a meta description:
  * The `DocumentHead` has been extracted into a component that can use `useTranslate`. The previous static `translate` call did not work.
  * i18n of the content in `renderHeaderSection` has been fixed in a similar way.
  * A localized meta description has been added to the extracted `DocumentHead` to improve SEO.

* Enables SSR on `/discover` to render the meta description and other parts for SEO purposes:
  * SSR has been enabled by setting `isomorphic: true`, adding a `package.json`, and splitting out `index.web.js` and `index.node.js`.
  * In `discover-stream.js`, the `DiscoverHeader` component was extracted into a real component so that it can also be SSR'ed.
  * The `discover` function has been moved from `controller.js` to `index.web.js` (after which `controller.js` could be removes). `discoverSsr`, a stripped-down version of `discover`, has been added to `index.node.js`. It only renders the `DocumentHead`, the page's header, and a placeholder to indicate that posts are being loaded. To keep it light, the feed itself is not SSR'ed.

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

The search engine snippet is incorrectly localized and contains placeholder content. This PR fixes this, improving SEO. Here are the Dutch and English snippets:

|English|Dutch|
|-|-|
|<img width="616" alt="image" src="https://github.com/user-attachments/assets/f84f936a-f73d-4f1b-abfe-008e65d1f563">|<img width="653" alt="image" src="https://github.com/user-attachments/assets/8cd90127-eef0-4dec-b5d2-13762ed01184">|

Plus, the page and feed weren't localized correctly. The window title and displayed posts were shown in English:

|Before|After|
|-|-|
|![image](https://github.com/user-attachments/assets/fabce8dc-b58b-40a7-9355-6166f2f79ea6)|![image](https://github.com/user-attachments/assets/805d88cc-fb72-403b-9b08-7b116b03928e)|


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**While logged out:**

* Visit `/[locale]/discover` with a non-existing locale (e.g. `/zz/discover`) and verify that it redirects to the English `/discover`.
* Compare `/discover` to production's https://wordpress.com/discover and verify that the pages are identical.
* Open `/[locale]/discover` with an existing locale (e.g. `/es/discover`) and verify that the window title is now translated. (Compare to production's https://wordpress.com/es/discover, if you want.)
* Verify that the feed now shows posts in the locale of your choice. In production, it currently shows English posts.
* View the page's source and verify that the `<meta name="description" [...]` is rendered, as well as the translated "Discover" and "Explore popular blogs that inspire, educate, and entertain." texts.
* Browse around the page (click some tabs and posts) and verify that the feed works normally.

**While logged in:**
* Switch to a non-English WordPress.com UI.
* Visit `/discover`. Compare it to production's https://wordpress.com/discover and verify that the pages are identical.
* Visit `/[locale]/discover` with a locale that's different from your UI's.
* Verify that you're redirected to `/discover` and that you see the page in your UI's locale (instead of the locale you used in the URL). (I think this only works correctly on calypso.live, or maybe I need to set a certain flag locally.)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

jornp committed:
Reader Discover i18n: Enable SSR and add a meta description

jornp committed:
Extract DiscoverHeader from AsyncLoad to SSR it

jornp committed:
Remove sidebar from SSR (only used when logged in)

jornp committed:
Move DiscoverHeader to discover-stream to render it inside the stream

jornp committed:
Add an SSR PostPlaceholder and remove updateLastRoute

jornp committed:
Clean up code

jornp committed:
Check whether we have props.width

jornp committed:
Improve index.node.js

redirectInvalidLanguage is a no-op on the server. And simply match
anyLangParam to match the section definition.

jornp committed:
Improve index.web.js

Handle anyLangParam only (which also covers langParam) and move
redirectInvalidLanguage to the list of middlewares instead of placing it
in a separate handler.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/16882967

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @jornp for including a screenshot in the description! This is really helpful for our translators.

tyxla commented on client/reader/discover/controller.js:
```diff
@@ -73,8 +90,27 @@ const exported = {
+		}
+		const selectedTab = context.query.selectedTab || DEFAULT_TAB;
+		const tabTitle = getSelectedTabTitle( selectedTab ) || '';
+		/* eslint-disable wpcalypso/jsx-classname-namespace */
```
Feel free to get rid of those `eslint-disable` rules in this file. Seems like since #66137 the rule only checks for following the BEM convention.

jornp replied:
Good to know; I've removed the rules.

tyxla commented on client/reader/discover/controller.js:
```diff
@@ -73,8 +90,27 @@ const exported = {
+			context.renderHeaderSection = renderHeaderSection;
+		}
+		const selectedTab = context.query.selectedTab || DEFAULT_TAB;
+		const tabTitle = getSelectedTabTitle( selectedTab ) || '';
```
Unless I'm missing something, `getSelectedTabTitle( selectedTab )` will never be falsy (it will default to `selectedTab` which is never falsy), so no need to default to an empty string:

```suggestion
		const tabTitle = getSelectedTabTitle( selectedTab );
```

jornp replied:
Yeah, that's true. I've removed the `|| ''`s.

tyxla commented on client/reader/discover/controller.js:
```diff
@@ -73,8 +90,27 @@ const exported = {
+
+	discoverSsr( context, next ) {
+		const state = context.store.getState();
+		if ( ! isUserLoggedIn( state ) ) {
+			context.renderHeaderSection = renderHeaderSection;
+		}
```
SSR won't run for logged-in users, so this check should be unnecessary:

```suggestion
		context.renderHeaderSection = renderHeaderSection;
```

jornp replied:
Good point, fixed!

tyxla commented on client/reader/discover/controller.js:
```diff
@@ -73,8 +90,27 @@ const exported = {
 	},
+
+	discoverSsr( context, next ) {
+		const state = context.store.getState();
```
This line won't be needed anymore (see comments below):
```suggestion
```

tyxla commented on client/reader/discover/discover-stream.js:
```diff
@@ -23,6 +23,33 @@ import {
+
+	return (
+		<NavigationHeader
+			navigationItems={ [] }
```
Pre-existing, but while we're here: I'd suggest that we declare this array outside the component as a constant, and always pass the same array to `navigationItems`.

```suggestion
			navigationItems={ [] }
```

jornp replied:
Good point, thanks! I've moved it outside the component as a const.

tyxla commented on client/reader/discover/discover-stream.js:
```diff
@@ -127,16 +133,14 @@ const DiscoverStream = ( props ) => {
-			</Stream>
-		</>
+		<Stream { ...streamProps }>
+			<DiscoverHeader selectedTab={ selectedTab } width={ props.width } />
```
Thanks for fixing this usage 👍

tyxla commented on client/reader/discover/discover-stream.js:
```diff
@@ -23,6 +23,33 @@ import {
+			title={ translate( 'Discover' ) }
+			subtitle={ subHeaderText }
+			className={ clsx( 'discover-stream-header', {
+				'reader-dual-column': Number.isInteger( props.width ) && props.width > WIDE_DISPLAY_CUTOFF,
```
Is the `Number.isInteger` check really needed? It will work just fine if `props.width` is `undefined` or `null`. 

```suggestion
				'reader-dual-column': Number.isInteger( props.width ) && props.width > WIDE_DISPLAY_CUTOFF,
```

jornp replied:
Thank you! I've removed the check.

tyxla commented on client/reader/lib/header-section/index.tsx:
```diff
@@ -1,16 +1,22 @@
+					translate( 'WordPress Reader' )
+				}
+			</h2>
+			<h1>{ translate( 'Enjoy millions of blogs at your fingertips.' ) }</h1>
```
Pre-existing, but I found it an odd semantic decision to have the description/subtitle in a `h1` and the title in a `h2` 🤔

jornp replied:
Yeah, that does feel unusual. Still, I could also get behind the reasoning that WordPress Reader is more like a "branding" and not necessarily the best description of the page itself.

I can check on this within MarTech to see what's most appropriate! We can always follow up on this.

tyxla replied:
Yep, definitely not a blocker for this PR 👍

jsnajdr replied:
The rendered page looks like this, the hiearchy with `h2` above `h1` seems justified:
<img width="1003" alt="Screenshot 2024-09-30 at 15 27 50" src="https://github.com/user-attachments/assets/6165a6e1-920b-41ef-ae84-f43e635330da">

tyxla commented on client/reader/lib/header-section/index.tsx:
```diff
@@ -1,16 +1,22 @@
+	);
+};
+
+const renderHeaderSection = () => <ReaderHeaderSection />;
 
 export default renderHeaderSection;
```
Is this indirection necessary? Can we just export `ReaderHeaderSection`?

```suggestion
export default ReaderHeaderSection;
```

jornp replied:
Good point. I don't think the indirection is strictly necessary, but I noticed that the exported value is exclusively used as `context.renderHeaderSection = renderHeaderSection;` across the Reader. I thought that by exporting `ReaderHeaderSection` directly, we might lose the context that it's meant to be assigned to `context.renderHeaderSection`. But we could add that as a comment, of course.

Just in case the `ReaderHeaderSection` name resembled `renderHeaderSection` too much, I renamed the component to `ReaderHeader`.

If my reasoning is unnecessary and we'd best just export `ReaderHeaderSection`, let me know!

tyxla replied:
Feel free to leave it as-is, but without further research, I expect this has something to do with ancient CJS code remnants and doesn't have a particular purpose. It is not a blocker for the PR by any means.

jornp replied:
Thanks, that makes sense! I'll remove the indirection in a separate PR, as it would probably be best to rename the related imports too, which would require changes to a couple of other Reader sub-packages and add a bit of noise to this PR.

tyxla commented on client/reader/discover/discover-stream.js:
```diff
@@ -127,16 +133,14 @@ const DiscoverStream = ( props ) => {
-		</>
+		<Stream { ...streamProps }>
+			<DiscoverHeader selectedTab={ selectedTab } width={ props.width } />
+			<DiscoverNavigation
```
I assume it's expected that on SSR this won't load because we're doing this for the search engines?

![Screenshot 2024-09-27 at 18 35 23](https://github.com/user-attachments/assets/d5464c52-53aa-42e5-82d2-ecea9c4a3b08)

jornp replied:
Yes, this is how we intended it. The main goal of enabling SSR is indeed to improve SEO, by rendering the localized meta description and the `Descubre` and `Explora [...]` texts. Other than that, we wanted to keep the SSR light and not render the feed itself.

We thought showing this `PostPlaceholder` would be a small, user-friendly touch to indicate that content will be loaded there.

Do you think this is a good UX or would you change it? Thanks!

tyxla replied:
Sounds good for this pass 👍 In combination with the "please enable JS" popup, it should be enough.

tyxla commented on client/reader/discover/index.web.js:
```diff
@@ -10,13 +9,12 @@ import { setLocaleMiddleware } from 'calypso/controller/shared';
-	page(
-		[ '/discover', `/${ langParam }/discover` ],
+	router(
+		[ '/discover', `/${ anyLangParam }/discover` ],
```
It's a bit unfortunate that this jumps so much when testing with a logged-in user 😕 


https://github.com/user-attachments/assets/17fef59e-98aa-45a3-ae11-4abd9070a612

jornp replied:
Thanks for this, I agree. I assumed the jump to and from the logged-out layout only occurred locally, but I now see it happens on calypso.live too, and it's probably caused by the changes in this PR. That's not how I intended it so that should be fixed!

yuliyan replied:
I believe calypso.live instances have different behavior when it comes to setting the logged-in state on the server side compared to production/staging running on `wordpress.com`.

When booting Calypso locally with `wpcom-user-bootstrap` enabled, the SSR flashing no longer happens, which makes me believe that this issue won't occur on production.

yuliyan replied:
It appears that we have the same SSR flashing on the `/themes` section in calypso.live and on production, it skips the SSR as intended.

jornp replied:
Thanks for checking, @yuliyan. That makes sense. I just tried running it locally with `wpcom-user-bootstrap`, but I still see the flashing. From a quick look at the code, I think I need a `wordpress_logged_in_cookie` in my `config/secrets.json`, which I set, but no luck yet. Maybe I'm still not configured correctly.

tyxla replied:
FWIW, @yuliyan is right, I had forgotten about this behavior. 

You won't be able to test it locally, because SSR isn't disabled locally. 

You can verify this with `/themes` and `/plugins` on staging/production - it won't flash there because the user bootstrap works correctly and SSR is disabled. 

Feel free to disregard my initial comment, nothing to do here 👍

jornp committed:
Remove eslint-disable rules from controller.js

jornp committed:
Remove the `|| ''` after the getSelectedTabTitle calls

The result of getSelectedTabTitle is always a string.

jornp committed:
Remove the isUserLoggedIn check in discoverSsr

SSR will only run for logged out users.

jornp commented:
@tyxla Thank you so much for the detailed review, I appreciate it! I'm working on the improvements.

jornp committed:
Pass a constant navigationItems array to NavigationHeader

jornp committed:
Remove the Number.isInteger check on props.width in DiscoverHeader

It'll work without the check, even when it's not passed to the
<DiscoverHeader> rendered from SSR.

jornp committed:
Rename ReaderHeaderSection to ReaderHeader for clarity

tyxla reviewed (approved):
Thanks for addressing all the feedback @jornp!

This looks good to go from my perspective 🚀 

As always, let's monitor server dashboards after deploying this.

yuliyan commented on client/reader/discover/controller.js:
```diff
@@ -8,14 +8,36 @@ import {
+	const translate = useTranslate();
+
+	const title = translate( 'Browse %s blogs & read articles ‹ Reader', {
+		args: [ tabTitle ],
```
Not related to the changes in this PR, but I just noticed that the `tabTitle` argument will have [a hardcoded untranslated value](https://github.com/Automattic/wp-calypso/blob/c9467f7afcc777c2850861e67f9c8bf10cdbf2c5/client/reader/discover/helper.js#L25).

jornp replied:
Thanks, I saw that too! We'll follow up on this.

yuliyan commented on client/reader/discover/controller.js:
```diff
@@ -70,11 +86,25 @@ const exported = {
+	discoverSsr( context, next ) {
+		context.renderHeaderSection = renderHeaderSection;
+
+		const selectedTab = context.query.selectedTab || DEFAULT_TAB;
```
[The SSR is skipped when there's a query strings](https://github.com/Automattic/wp-calypso/blob/c9467f7afcc777c2850861e67f9c8bf10cdbf2c5/client/server/render/index.js#L304), but it should be fine to leave it as is, in case this behavior is updated in the future.

jornp replied:
Thanks for the link to the source! I did notice the behavior too. I also think it's fine as-is — with the current changes, at least the main Discover page will be indexed nicely, which will probably have the biggest impact.

yuliyan reviewed (approved):
The changes look good and work as expected! 🚀

jsnajdr commented on client/reader/discover/controller.js:
```diff
@@ -70,11 +86,25 @@ const exported = {
+		next();
+	},
+
+	discoverSsr( context, next ) {
```
You might want to put the `discover` and `discoverSsr` functions into separate files. Otherwise the browser bundle will include dead code for SSR and node bundle will include unneccesary code that's active only in the browser.

Here there is probably not so much code like this. I only see the `track*` functions from `controller-helper`. But it's still a good idea to keep them separate.

jornp replied:
Thanks, good point! I've moved the functions to their respective index file.

jsnajdr commented on client/reader/discover/controller.js:
```diff
@@ -70,11 +86,25 @@ const exported = {
 	},
 };
 
 export default exported;
```
Could we get rid of this `exported` object and simply export the `discover*` functions as plain ESM exports? The object is probably some remnant of old CommonJS modules.

jornp committed:
Move `discover` and `discoverSsr` to index.web.js and index.node.js

And extract DiscoverDocumentHead to a separate file.

jornp commented:
@jsnajdr Thank you for the review! Those are good points.

I've moved `discover` to `index.web.js` and `discoverSsr` to `index.node.js`. Plus, I've extracted the `DiscoverDocumentHead` into its own file (`discover-document-head.js`).

I hope this is better!

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/16882967).

yuliyan reviewed (approved):
The file structure changes look good to me. Confirming that the PR still works as expected. 🚀
